geode-junit/src/main/java/org/apache/geode/test/junit/rules/ConcurrencyRule.java

```diff
@@ -381,9 +408,16 @@ private void callAndValidate() throws Exception {
       } else if (expectedException != null) {
         Throwable thrown = catchThrowable(() -> this.callable.call());
         checkThrown(this.expectedException, thrown);
+      } else if (expectedExceptionType != null && expectedExceptionCauseType != null) {
+        Throwable thrown = catchThrowable(() -> this.callable.call());
+        assertThat(thrown).isInstanceOf(expectedExceptionType)
+            .hasCauseInstanceOf(expectedExceptionCauseType);
       } else if (expectedExceptionType != null) {
         Throwable thrown = catchThrowable(() -> this.callable.call());
         assertThat(thrown).isInstanceOf(this.expectedExceptionType);
+      } else if (expectedExceptionCauseType != null) {
+        Throwable thrown = catchThrowable(() -> this.callable.call());
+        assertThat(thrown.getCause()).isInstanceOf(this.expectedExceptionCauseType);
```

Comment:
I would refactor this method to have the `this.callable.call()` in a try/catch. In the catch block you can test for the various expectedException combinations.
```
      T result = null;
      try {
        result = this.callable.call();
        if (expectedValue != null) {
          assertThat(result).isEqualTo(this.expectedValue);
        }
      } catch(Throwable thrown) {
        if (expectedException != null) {
          ...
        } else if (expectedExceptionType != null && expectedExceptionCauseType != null) {
          ...
        } else if (expectedExceptionType != null) {
          ...
        } else if (expectedExceptionCauseType != null) {
          ...
        } else {
          throw thrown;
        }
      }
```

Reply:
I did write that exact implementation at one point then deleted it because it didn't seem like much of an improvement on readability. But maybe it is worth doing anyway just because it makes more sense to repeat less code